source/data-formats/codecs.txt

```diff
@@ -0,0 +1,390 @@
+
+BsonTypeClassMap
+~~~~~~~~~~~~~~~~
+
+The ``BsonTypeClassMap`` class contains a recommended mapping between BSON
+and {+language+} types. You can use this class in your custom ``Codec`` or
+``CodecProvider`` to help you manage which {+language+} types to decode your BSON
+types to. It also contains container classes that implement ``Iterable``
+or ``Map`` such as the ``Document`` class.
+
+You can add or modify the ``BsonTypeClassMap`` default mapping by passing a
+``Map`` containing new or replacement entries.
+
+The following code shows how to retrieve the {+language+} class type
+that corresponds to the BSON array type in the default ``BsonTypeClassMap``
+instance:
+
+.. io-code-block::
+
+   .. input::
+      :language: kotlin
+
+      val bsonTypeClassMap = BsonTypeClassMap()
+      val clazz = bsonTypeClassMap[BsonType.ARRAY]
+      println("Kotlin class name: " + clazz.name)
+
+   .. output::
+      :language: console
+
+      Kotlin class name: java.util.List
+
+You can modify these mappings in your instance by specifying replacements in the
+``BsonTypeClassMap`` constructor. The following code snippet shows how
+you can replace the mapping for the BSON array type in your ``BsonTypeClassMap``
+instance with the ``Set`` class:
+
+.. io-code-block::
+
+   .. input::
+      :language: kotlin
+
+      val replacements = mutableMapOf<BsonType, Class<*>>(BsonType.ARRAY to MutableSet::class.java)
+      val bsonTypeClassMap = BsonTypeClassMap(replacements)
+      val clazz = bsonTypeClassMap[BsonType.ARRAY]
+      println("Class name: " + clazz.name)
+
+   .. output::
+      :language: console
+
+      Kotlin class name: java.util.Set
+
+For a complete list of the default mappings, view the
+`BsonTypeClassMap
+<{+java-api+}/apidocs/bson/org/bson/codecs/BsonTypeClassMap.html>`__ API
+documentation.
+
+.. _kotlin-sync-codec-custom-example:
+
+Custom Codec Example
+--------------------
+
+This section demonstrates how you can implement ``Codec`` and
+``CodecProvider`` interfaces to define the encoding and decoding logic
+for a custom {+language+} class. It shows how you can specify and use
+your custom implementations to perform read and write operations.
+
+The following code defines the sample data class ``Monolight``:
+
+.. literalinclude:: /includes/data-formats/codecs.kt
+   :language: kotlin
+   :start-after: start-monolight-class
+   :end-before: end-monolight-class
+   :dedent:
+
+This class contains the following fields that each require a
+corresponding ``Codec`` to handle encoding and decoding:
+
+- ``powerStatus``: Describes whether the device light is ``"ON"`` or ``"OFF"``.
+  For this field, use the :ref:`PowerStatusCodec
+  <kotlin-sync-powerstatus-codec>` that converts the ``PowerStatus``
+  enum values to BSON booleans.
+
+- ``colorTemperature``: Describes the color of the device light in
+  Kelvins as an ``Int`` value. For this field, use the ``IntegerCodec``
```

Comment:
[nit] I _think_ this is usually singular if capitalized, or lowercase when plural: 
```suggestion
  kelvins as an ``Int`` value. For this field, use the ``IntegerCodec``
```
or 
```suggestion
  Kelvin as an ``Int`` value. For this field, use the ``IntegerCodec``
```